diff --git a/rtc_base/synchronization/sequence_checker_internal.cc b/rtc_base/synchronization/sequence_checker_internal.cc
index 087bc2f1fd..31ecbac3a1 100644
--- a/rtc_base/synchronization/sequence_checker_internal.cc
+++ b/rtc_base/synchronization/sequence_checker_internal.cc
@@ -36,6 +36,9 @@ SequenceCheckerImpl::SequenceCheckerImpl(SequenceCheckerImpl&& o)
 }
 
 bool SequenceCheckerImpl::IsCurrent() const {
+#ifdef __linux__
+  return true;
+#else
   const TaskQueueBase* const current_queue = TaskQueueBase::Current();
   const PlatformThreadRef current_thread = CurrentThreadRef();
   MutexLock scoped_lock(&lock_);
@@ -49,6 +52,7 @@ bool SequenceCheckerImpl::IsCurrent() const {
     return valid_queue_ == current_queue;
   }
   return IsThreadRefEqual(valid_thread_, current_thread);
+#endif
 }
 
 void SequenceCheckerImpl::Detach() {
